core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -875,9 +888,10 @@ class ReplicaManager(val config: KafkaConfig,
           fetchPartitionStatus += (topicPartition -> FetchPartitionStatus(logOffsetMetadata, partitionData))
         })
       }
-      val fetchMetadata = FetchMetadata(fetchMinBytes, fetchMaxBytes, hardMaxBytesLimit, fetchOnlyFromLeader,
-        fetchIsolation, isFromFollower, replicaId, fetchPartitionStatus)
-      val delayedFetch = new DelayedFetch(timeout, fetchMetadata, this, quota, responseCallback)
+      val fetchMetadata = FetchMetadata(fetchMinBytes, fetchMaxBytes, hardMaxBytesLimit, isFromFollower,
+        fetchIsolation, isFromFollower, replicaId, fetchPartitionStatus, hasFetchSession)
+      val delayedFetch = new DelayedFetch(timeout, fetchMetadata, this, quota, clientMetadata,
```

Comment:
If the follower has no cached high watermark, we want to respond immediately. Currently we have this logic in DelayedFetch, but I think it makes more sense to avoid purgatory altogether and respond immediately after the initial fetch above in this function.